Split CI workflows into 'test' and 'deploy'

nicoddemus opened the pull request:
This makes them easier to maintain and follows how other pytest repositories have evolved.

Also made some changes to the deploy workflow:

* Using Python 3.10
* Using `build` package instead of calling `python setup.py`

nicoddemus committed:
Split CI workflows into 'test' and 'deploy'

This makes them easier to maintain and follows how other pytest repositories have evolved.

nicoddemus commented on .github/workflows/deploy.yml:
```diff
@@ -0,0 +1,36 @@
+    tags:
+      - "*"
+
+permissions: {}
```
Copied from pytest

RonnyPfannschmidt commented:
im a bit unhappy with the splitting, i'd prefer if we started to install from the build artifacts for testing (like i have done for some of my projects)

nicoddemus commented:
> im a bit unhappy with the splitting, i'd prefer if we started to install from the build artifacts for testing (like i have done for some of my projects)

What do you mean? IIUC, you mean to test the package instead of source, but isn't that what `tox` does for us?

Or do you mean to generate a single artifact, test over it, and then deploy that same artifact?

nicoddemus commented:
Happy to try something like that out, can you point to a project to use as reference?

RonnyPfannschmidt commented:
@nicoddemus i'd love to extract this into something reusable, but currently im a bit tied up with other priorities

bluetech committed:
ci: tighten job permissions

bluetech committed:
ci: don't persist credentials from checkout

Not needed, improves security.

bluetech committed:
ci: enable github build provenance attestation

See https://github.com/hynek/build-and-inspect-python-package#inputs
Following pytest.

bluetech committed:
dependabot: add 7 day cooldown period

As suggested by zizmor.

https://docs.zizmor.sh/audits/#dependabot-cooldown

bluetech committed:
ci: pin github actions

Fix https://docs.zizmor.sh/audits/#unpinned-uses.

Used `uvx gha-update`: https://github.com/davidism/gha-update

bluetech committed:
ci: fix zizmor template-injection suggestions

Not actually vulnerable, but better use the safe pattern anyway.

https://docs.zizmor.sh/audits/#template-injection

bluetech commented:
Rebased and also brought over some other improvements from pytest, and some zizmor fixes.
